packages/shared/src/hooks/useChat.ts

```diff
@@ -21,130 +27,136 @@ enum UseChatMessageType {
+              idQueryKey,
+              initializeSearchSession({
+                ...(data.payload as CreatePayload),
+                createdAt: new Date(),
+                status: data.status,
+                prompt,
+              }),
+            );
             break;
-          }
           case UseChatMessageType.WebSearchFinished:
-          case UseChatMessageType.WebResultsFiltered:
-          case UseChatMessageType.StatusUpdated: {
-            setState((current) => ({
-              ...current,
+            setSearchQuery({
+              sources: (data.payload as SourcesMessage).sources,
               status: data.status,
-            }));
-
+            });
             break;
-          }
-          case UseChatMessageType.NewTokenReceived: {
-            const { token } = data.payload as { token: string };
-            // TODO WT-1554-stream-rendering handle sources
-
-            setMessages((current) => [(current[0] || '') + token]);
+          case UseChatMessageType.WebResultsFiltered:
+            setSearchQuery({ status: data.status });
+            break;
+          case UseChatMessageType.StatusUpdated:
+            setSearchQuery({ status: data.status });
+            break;
+          case UseChatMessageType.NewTokenReceived:
+            setSearchQuery({ response: (data.payload as TokenPayload).token });
             break;
-          }
           case UseChatMessageType.Completed: {
-            setState((current) => ({
-              ...current,
-              isLoading: false,
-            }));
+            setSearchQuery({ completedAt: new Date(), status: data.status });
             setPrompt(undefined);
-
             break;
           }
-          case UseChatMessageType.Error: {
-            setState((current) => ({
-              ...current,
-              error: data.payload as UseChat['error'],
-            }));
-
+          case UseChatMessageType.Error:
+            setSearchQuery({ error: data.payload as SearchChunkError });
             break;
-          }
-          case UseChatMessageType.SessionFound: {
-            // TODO WT-1554-stream-rendering redirect to session page
-
+          case UseChatMessageType.SessionFound:
+            client.setQueryData(idQueryKey, () => data.payload as Search);
             break;
-          }
           default:
             break;
         }
       } catch (error) {
         // eslint-disable-next-line no-console
         console.error('[EventSource][message] error', error);
       }
-    };
+    },
+    [client, prompt, idQueryKey, setSearchQuery],
+  );
 
-    // eslint-disable-next-line @typescript-eslint/no-use-before-define
-    const endStream = mockChatStream({ onMessage });
+  const executePrompt = useCallback(
+    (value: string) => {
+      if (!value) {
+        return;
+      }
 
-    return () => {
-      endStream();
-    };
-  }, [prompt]);
+      if (endStreamRef.current) {
+        endStreamRef.current();
+      }
```

Comment:
My only concern with this approach If hook unmounts the stream will be left open.

Reply:
We can probably clean it in empty useEffect with only cleanup function.

Reply:
Oh, right. Yes, let's open a `useEffect` to clean it up.